### DISCUSSION TITLE
Print diagnostics with non-owned source code

### DISCUSSION THREAD

**Discussion by MattX** [zkat/miette#391] [Q&A] [answered]
Hello! I'm writing a small interpreter for a DSL, and would like to use miette for error reporting,

I've tried defining an error type like this:
```
#[derive(Debug, Error, Diagnostic)]
#[error("Runtime error")]
pub enum RuntimeError<'source> {
    #[error("Empty input")]
    EmptyInput {
        #[source_code]
        source_code: &'source str,

        #[label = "Empty input"]
        range: SourceSpan,
    },
    // ...
}
```

But calling `.into_diagnostic()` on something that returns `Result<T, RuntimeError<'source>>` [creates a `'source = 'static` requirement](https://docs.rs/miette/latest/miette/trait.IntoDiagnostic.html#foreign-impls), whereas I would like my source code to be temporary (in fact I want to drop it immediately after printing the diagnostic).

I know I could wrap the source in an `Rc` or use another trick to extend its lifetime, but is there a way to reduce the lifetime requirement to "alive while the diagnostic is being printed"?

**Comment by zkat** [comment]
You can "delay" passing the source code in (and thus make it temporary) by using [`Report::with_source_code`](https://docs.rs/miette/latest/miette/#-delayed-source-code)

**Comment by MattX** [comment]
But that still requires the source code to be `'static`, doesn't it? That trait bound exists in [`Report::with_source_code`](https://docs.rs/miette/latest/miette/struct.Error.html#method.with_source_code). It feels like it should be possible to create a `Report` with a non-`'static` source code, as long as the `Report` doesn't outlive the source code?

**Comment by scriptandcompile** [comment]
I would also like some assistance with this. I'm running into basically the same issue. I can just clone it, but since I was counting on my error type to extremely light weight. I'm using it as the error side of a parser combinator and so returning the error side is *common*. I've noticed a *massive* slow down when I tried to integrate source code reporting to my error reporting.

**Comment by MattX** [comment, marked as answer]
In the end I've used a wrapper type that implements `Diagnostic` and delegates every method to a wrapped `Diagnostic`, except `source_code()`, which it gets from another field. Near the top-level of my code, I wrap the error with this type and bubble that up instead.

It's just a lot of boilerplate: https://gist.github.com/MattX/e6e01b453aad4778947e54a58c6f5be5
